File: examples/diagnostic-showcase/rust-cli/src/example_chained.rs
use std::fmt::{self, Display, Formatter};
use std::ptr::NonNull;
use std::rc::Rc;

use miette::{
    Diagnostic, LabeledSpan, MietteSpanContents, Severity, SourceCode, SourceSpan,
};

#[derive(Clone)]
pub struct Source {
    pub name: String,
    pub data: String,
}

impl SourceCode for Source {
    fn read_span<'a>(
        &'a self,
        span: &SourceSpan,
        _before: usize,
        _after: usize,
    ) -> Result<Box<dyn miette::SpanContents<'a> + 'a>, miette::MietteError> {
        let mut line = 1;
        let mut column = 1;
        for b in self.data.as_bytes()[..self.data.len().min(span.offset())].iter() {
            if *b == b'\n' {
                line += 1;
                column = 1;
            } else {
                column += 1;
            }
        }

        let mut line_count = 1;
        for b in self.data.as_bytes()
            [span.offset()..self.data.len().min(span.offset() + span.len())]
            .iter()
        {
            if *b == b'\n' {
                line_count += 1;
            }
        }

        Ok(Box::new(MietteSpanContents::new_named(
            self.name.clone(),
            self.data.as_bytes(),
            *span,
            line,
            column,
            line_count,
        )))
    }
}

#[derive(Clone, Copy)]
struct Label {
    text: Option<&'static str>,
    offset: usize,
    len: usize,
    primary: bool,
}

pub struct ChainedDiagnostic {
    message: &'static str,
    code: Option<&'static str>,
    severity: Option<Severity>,
    help: Option<&'static str>,
    url: Option<&'static str>,
    source_idx: Option<usize>,
    sources: Rc<Vec<Source>>,
    labels: &'static [Label],
    related: Vec<*const ChainedDiagnostic>,
    diag_source: Option<*const ChainedDiagnostic>,
}

// SAFETY: ChainedDiagnostic only contains raw pointers to other
// ChainedDiagnostic values within the same stable Vec allocation.
// No ownership or mutation through these pointers.
unsafe impl Send for ChainedDiagnostic {}
unsafe impl Sync for ChainedDiagnostic {}

impl ChainedDiagnostic {
    fn source(&self) -> Option<&Source> {
        self.source_idx.map(|i| &self.sources[i])
    }
}

impl Display for ChainedDiagnostic {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        f.write_str(self.message)
    }
}

impl fmt::Debug for ChainedDiagnostic {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        f.debug_struct("ChainedDiagnostic")
            .field("message", &self.message)
            .finish()
    }
}

impl std::error::Error for ChainedDiagnostic {}

impl Diagnostic for ChainedDiagnostic {
    fn code<'a>(&'a self) -> Option<Box<dyn Display + 'a>> {
        self.code.map(|c| Box::new(c) as Box<dyn Display>)
    }

    fn severity(&self) -> Option<Severity> {
        self.severity
    }

    fn help<'a>(&'a self) -> Option<Box<dyn Display + 'a>> {
        self.help.map(|h| Box::new(h) as Box<dyn Display>)
    }

    fn url<'a>(&'a self) -> Option<Box<dyn Display + 'a>> {
        self.url.map(|u| Box::new(u) as Box<dyn Display>)
    }

    fn source_code(&self) -> Option<&dyn SourceCode> {
        self.source().map(|s| s as &dyn SourceCode)
    }

    fn labels(&self) -> Option<Box<dyn Iterator<Item = LabeledSpan> + '_>> {
        if self.labels.is_empty() {
            return None;
        }
        Some(Box::new(self.labels.iter().map(|label| {
            let span: SourceSpan = (label.offset, label.len).into();
            if label.primary {
                LabeledSpan::new_primary_with_span(label.text.map(str::to_owned), span)
            } else {
                LabeledSpan::new_with_span(label.text.map(str::to_owned), span)
            }
        })))
    }

    fn related<'a>(&'a self) -> Option<Box<dyn Iterator<Item = &'a dyn Diagnostic> + 'a>> {
        if self.related.is_empty() {
            return None;
        }
        // SAFETY: all raw pointers point to ChainedDiagnostic values
        // within the same stable Vec that outlives &self.
        let refs: Vec<&'a dyn Diagnostic> = self
            .related
            .iter()
            .map(|&p| unsafe { &*p as &dyn Diagnostic })
            .collect();
        Some(Box::new(refs.into_iter()))
    }

    fn diagnostic_source(&self) -> Option<&dyn Diagnostic> {
        // SAFETY: same lifetime invariant as related()
        self.diag_source
            .map(|p| unsafe { &*p as &dyn Diagnostic })
    }
}

struct NodeDesc {
    message: &'static str,
    code: Option<&'static str>,
    severity: Option<Severity>,
    help: Option<&'static str>,
    url: Option<&'static str>,
    source_idx: Option<usize>,
    labels: &'static [Label],
    related_idxs: &'static [usize],
    diag_source_idx: Option<usize>,
}

const fn primary(text: &'static str, offset: usize, len: usize) -> Label {
    Label { text: Some(text), offset, len, primary: true }
}
const fn secondary(text: &'static str, offset: usize, len: usize) -> Label {
    Label { text: Some(text), offset, len, primary: false }
}

const DESCS: &[NodeDesc] = &[
    // 00: standalone
    NodeDesc {
        message: "standalone diagnostic with no chain",
        code: Some("S000"),
        severity: Some(Severity::Advice),
        help: Some("everything is fine"),
        url: None,
        source_idx: None,
        labels: &[],
        related_idxs: &[],
        diag_source_idx: None,
    },
    // 01: single related sibling
    NodeDesc {
        message: "single related sibling",
        code: Some("C001"),
        severity: Some(Severity::Error),
        help: Some("check the sibling diagnostic for context"),
        url: None,
        source_idx: None,
        labels: &[],
        related_idxs: &[0],
        diag_source_idx: None,
    },
    // 02: multiple related siblings (3, with repetition)
    NodeDesc {
        message: "multiple related siblings",
        code: Some("C002"),
        severity: Some(Severity::Warning),
        help: Some("all siblings must be resolved"),
        url: None,
        source_idx: None,
        labels: &[],
        related_idxs: &[0, 1, 0],
        diag_source_idx: None,
    },
    // 03: causal chain depth 2
    NodeDesc {
        message: "causal chain depth 2 (d03 → d00)",
        code: Some("C003"),
        severity: Some(Severity::Error),
        help: Some("root cause is the standalone diagnostic"),
        url: None,
        source_idx: Some(2),
        labels: &[primary("ROW_NUMBER", 68, 10)],
        related_idxs: &[],
        diag_source_idx: Some(0),
    },
    // 04: causal chain depth 3
    NodeDesc {
        message: "causal chain depth 3 (d04 → d03 → d00)",
        code: Some("C004"),
        severity: Some(Severity::Error),
        help: Some("each step in the chain adds context"),
        url: None,
        source_idx: Some(3),
        labels: &[primary("TransformError", 479, 14)],
        related_idxs: &[],
        diag_source_idx: Some(3),
    },
    // 05: causal chain depth 4
    NodeDesc {
        message: "causal chain depth 4 (d05 → d04 → d03 → d00)",
        code: Some("C005"),
        severity: Some(Severity::Error),
        help: Some("deep causal chains reveal propagation path"),
        url: None,
        source_idx: Some(3),
        labels: &[primary("Semaphore", 269, 9)],
        related_idxs: &[],
        diag_source_idx: Some(4),
    },
    // 06: related with source labels
    NodeDesc {
        message: "related with source labels",
        code: Some("C006"),
        severity: Some(Severity::Warning),
        help: Some("related errors provide additional context"),
        url: None,
        source_idx: Some(4),
        labels: &[primary("title", 74, 5), secondary("items", 107, 5)],
        related_idxs: &[0, 2],
        diag_source_idx: None,
    },
    // 07: causal with source labels
    NodeDesc {
        message: "causal with source labels",
        code: Some("C007"),
        severity: Some(Severity::Error),
        help: Some("check the config values"),
        url: None,
        source_idx: Some(0),
        labels: &[primary("host", 10, 4), secondary("8080", 36, 4)],
        related_idxs: &[],
        diag_source_idx: Some(1),
    },
    // 08: both chain types
    NodeDesc {
        message: "both chain types (causal + related)",
        code: Some("C008"),
        severity: Some(Severity::Error),
        help: Some("causal chain explains why, related shows what else"),
        url: None,
        source_idx: Some(1),
        labels: &[primary("HealthCheck", 79, 11)],
        related_idxs: &[2, 3],
        diag_source_idx: Some(0),
    },
    // 09: nested related (d09 → d10 → d00 via related)
    NodeDesc {
        message: "nested related chain",
        code: Some("C009"),
        severity: Some(Severity::Error),
        help: Some("nested chains expand recursively"),
        url: None,
        source_idx: Some(2),
        labels: &[primary("ranked AS", 5, 9)],
        related_idxs: &[10],
        diag_source_idx: None,
    },
    // 10: nested inner
    NodeDesc {
        message: "nested related inner",
        code: Some("C009"),
        severity: Some(Severity::Warning),
        help: None,
        url: None,
        source_idx: Some(3),
        labels: &[secondary("worker", 306, 6)],
        related_idxs: &[0],
        diag_source_idx: None,
    },
    // 11: causal + related + labels
    NodeDesc {
        message: "causal plus related with labels",
        code: Some("C010"),
        severity: Some(Severity::Error),
        help: None,
        url: None,
        source_idx: Some(4),
        labels: &[primary("for item", 175, 8)],
        related_idxs: &[2, 4],
        diag_source_idx: Some(9),
    },
    // 12: diamond
    NodeDesc {
        message: "diamond: one cause with two related effects",
        code: Some("C011"),
        severity: Some(Severity::Error),
        help: Some("single root cause branches to multiple downstream errors"),
        url: None,
        source_idx: Some(3),
        labels: &[primary("asyncio.gather", 645, 14)],
        related_idxs: &[3, 4],
        diag_source_idx: Some(2),
    },
    // 13: cross-file
    NodeDesc {
        message: "cross-file chain spanning all sources",
        code: Some("C012"),
        severity: Some(Severity::Error),
        help: Some("related errors originate from five different files"),
        url: None,
        source_idx: None,
        labels: &[],
        related_idxs: &[0, 7, 3, 10, 12],
        diag_source_idx: None,
    },
    // 14: sub-labels
    NodeDesc {
        message: "related with sub-labels",
        code: Some("C013"),
        severity: Some(Severity::Warning),
        help: Some("each related has its own label within the same source"),
        url: None,
        source_idx: None,
        labels: &[],
        related_idxs: &[15, 16],
        diag_source_idx: None,
    },
    // 15: sub-label a
    NodeDesc {
        message: "ping context deadline exceeded",
        code: Some("C013"),
        severity: Some(Severity::Error),
        help: None,
        url: None,
        source_idx: Some(1),
        labels: &[primary("PingContext", 273, 11)],
        related_idxs: &[],
        diag_source_idx: None,
    },
    // 16: sub-label b
    NodeDesc {
        message: "database connection is nil",
        code: Some("C013"),
        severity: Some(Severity::Error),
        help: None,
        url: None,
        source_idx: Some(1),
        labels: &[primary("database not connected", 140, 22)],
        related_idxs: &[],
        diag_source_idx: None,
    },
    // 17: fan-out
    NodeDesc {
        message: "fan-out: one cause with multiple effects",
        code: Some("C014"),
        severity: Some(Severity::Error),
        help: Some("single failure triggers cascading errors"),
        url: None,
        source_idx: Some(0),
        labels: &[primary("pool_size", 108, 9)],
        related_idxs: &[7, 18, 19],
        diag_source_idx: Some(11),
    },
    // 18: fan-out sub a
    NodeDesc {
        message: "timeout connecting to database",
        code: Some("C014"),
        severity: Some(Severity::Error),
        help: None,
        url: None,
        source_idx: Some(1),
        labels: &[primary("fmt.Errorf", 128, 10)],
        related_idxs: &[],
        diag_source_idx: Some(1),
    },
    // 19: fan-out sub b
    NodeDesc {
        message: "transform pipeline rejected item",
        code: Some("C014"),
        severity: Some(Severity::Error),
        help: None,
        url: None,
        source_idx: Some(3),
        labels: &[primary("log.error", 516, 9)],
        related_idxs: &[],
        diag_source_idx: Some(2),
    },
    // 20: full chaos
    NodeDesc {
        message: "full chaos: causal + related + labels + cross-file + nested",
        code: Some("C019"),
        severity: Some(Severity::Error),
        help: Some("maximal diagnostic chain complexity"),
        url: Some("https://example.invalid/chaos"),
        source_idx: Some(2),
        labels: &[primary("employees", 146, 9), secondary("dept_id", 55, 7)],
        related_idxs: &[4, 7, 12, 15],
        diag_source_idx: Some(0),
    },
];

pub fn build(sources: Vec<Source>) -> Vec<ChainedDiagnostic> {
    let sources = Rc::new(sources);

    // Pre-allocate so pointers remain stable
    let mut diags = Vec::with_capacity(DESCS.len());

    // First pass: create all entries with empty chain fields
    for desc in DESCS {
        diags.push(ChainedDiagnostic {
            message: desc.message,
            code: desc.code,
            severity: desc.severity,
            help: desc.help,
            url: desc.url,
            source_idx: desc.source_idx,
            sources: Rc::clone(&sources),
            labels: desc.labels,
            related: Vec::new(),
            diag_source: None,
        });
    }

    // Take raw pointers while we have no mutable borrows
    let ptrs: Vec<*const ChainedDiagnostic> = diags.iter().map(|d| d as *const _).collect();

    // SAFETY: raw pointers point into the stable Vec allocation.
    // Vec will not reallocate because capacity == length.
    // Each diag's chain fields are only set once, here, before any
    // user code can access them.
    for (i, desc) in DESCS.iter().enumerate() {
        diags[i].related = desc
            .related_idxs
            .iter()
            .map(|&idx| ptrs[idx])
            .collect();
        diags[i].diag_source = desc.diag_source_idx.map(|idx| ptrs[idx]);
    }

    diags
}

File: examples/diagnostic-showcase/zig-cli/src/example_chained.zig
const diagnostic = @import("kbdiagnostic");
const span = @import("kbdiagnostic").span;
const source = @import("kbdiagnostic").source;

const Node = struct {
    message: []const u8,
    code: ?[]const u8,
    severity: ?diagnostic.Severity,
    help: ?[]const u8,
    url: ?[]const u8,
    source_code: ?source.SourceCode,
    labels: []const span.LabeledSpan,
    related: ?[]const diagnostic.Diagnostic,
    diag_source: ?*const diagnostic.Diagnostic,

fn implBy(self: *const Node) diagnostic.Diagnostic {
    return .{ .ptr = self, .vtable = &VTable };
}

    fn codeFn(ptr: *const anyopaque) ?[]const u8 {
        return node(ptr).code;
    }
    fn severityFn(ptr: *const anyopaque) ?diagnostic.Severity {
        return node(ptr).severity;
    }
    fn helpFn(ptr: *const anyopaque) ?[]const u8 {
        return node(ptr).help;
    }
    fn urlFn(ptr: *const anyopaque) ?[]const u8 {
        return node(ptr).url;
    }
    fn sourceCodeFn(ptr: *const anyopaque) ?*const source.SourceCode {
        const n = node(ptr);
        if (n.source_code) |*sc| return sc;
        return null;
    }
    fn labelsFn(ptr: *const anyopaque) ?[]const span.LabeledSpan {
        const n = node(ptr);
        return if (n.labels.len > 0) n.labels else null;
    }
    fn relatedFn(ptr: *const anyopaque) ?[]const diagnostic.Diagnostic {
        return node(ptr).related;
    }
    fn diagnosticSourceFn(ptr: *const anyopaque) ?*const diagnostic.Diagnostic {
        return node(ptr).diag_source;
    }
    fn messageFn(ptr: *const anyopaque) []const u8 {
        return node(ptr).message;
    }

    fn node(ptr: *const anyopaque) *const Node {
        return @ptrCast(@alignCast(ptr));
    }

    const VTable = diagnostic.Diagnostic.VTable{
        .code = codeFn,
        .severity = severityFn,
        .help = helpFn,
        .url = urlFn,
        .sourceCode = sourceCodeFn,
        .labels = labelsFn,
        .related = relatedFn,
        .diagnosticSource = diagnosticSourceFn,
        .message = messageFn,
    };
};

const NodeDesc = struct {
    message: []const u8,
    code: ?[]const u8 = null,
    severity: ?diagnostic.Severity = null,
    help: ?[]const u8 = null,
    url: ?[]const u8 = null,
    source_idx: ?usize = null,
    labels: []const span.LabeledSpan = &.{},
    related_idxs: []const usize = &.{},
    diag_source_idx: ?usize = null,
};

fn prim(text: []const u8, offset: usize, len: usize) span.LabeledSpan {
    return span.LabeledSpan.newPrimary(text, offset, len);
}
fn sec(text: []const u8, offset: usize, len: usize) span.LabeledSpan {
    return span.LabeledSpan.new(text, offset, len);
}

const DESCS = [_]NodeDesc{
    // 00: standalone — no chain
    .{
        .message = "standalone diagnostic with no chain",
        .code = "S000",
        .severity = .Advice,
        .help = "everything is fine",
    },
    // 01: single related sibling
    .{
        .message = "single related sibling",
        .code = "C001",
        .severity = .Error,
        .help = "check the sibling diagnostic for context",
        .related_idxs = &.{0},
    },
    // 02: multiple related siblings (3, with repetition)
    .{
        .message = "multiple related siblings",
        .code = "C002",
        .severity = .Warning,
        .help = "all siblings must be resolved",
        .related_idxs = &.{ 0, 1, 0 },
    },
    // 03: causal chain depth 2 — source is d00
    .{
        .message = "causal chain depth 2 (d03 → d00)",
        .code = "C003",
        .severity = .Error,
        .help = "root cause is the standalone diagnostic",
        .source_idx = 2,
        .labels = &.{prim("ROW_NUMBER", 68, 10)},
        .diag_source_idx = 0,
    },
    // 04: causal chain depth 4 — d04 → d09 → d02 → d00
    // d09 is the "nested inner" which has diag_source=d01 (wait no, d09 doesn't have diag_source)
    // Let me re-think: d04.diag_source->d07, d07.diag_source->d02, d02.diag_source->d00

    // Actually, I realize d02 has no diag_source. Let me use a different chain.
    // d04.diag_source = d06, d06.diag_source = d10, d10.diag_source = d00
    // But d06 doesn't have diag_source either.

    // Let me define separate "chain node" that only serves as causal link:
    // d04.diag_source = d15, d15.diag_source = d16, d16.diag_source = d00
    // where d15 and d16 are "sub-label" nodes with diag_source set.

    // UPDATE: reordering. Let me use d03 (index 3) as second link and d00 as first.
    // d04.diag_source = &d03, d03.diag_source = &d00
    // That's depth 3 (d04→d03→d00). For depth 4 need one more: d04→d07→d03→d00

    // But d07 doesn't exist yet. Let me just do d04 → d03 → d00 (depth 3 shown as causal chain depth 3),
    // and rename the message.

    // No, I'll restructure the indices. Let me just define them carefully:
    // d04.diag_source = d03, d03.diag_source = d00 → depth 3
    // For depth 4: I'll use d05.diag_source = d04, d04.diag_source = d03, d03.diag_source = d00

    // But d05 has already been assigned a different meaning ("related with source labels").
    // I need to reorder. Let me reassign the entire set.

    // FINAL REORDERING (committing to this):
    // 00: standalone
    // 01: single related [00]
    // 02: multi related [00,01,00]
    // 03: causal-2: diag_source=00, src=query.sql
    // 04: causal-4: diag_source=03→00 (depth 3, renamed to "causal chain depth 3")
    // 05: causal-5: diag_source=04→03→00 (depth 4, renamed to "causal chain depth 4")
    // 06: related+w/labels: related=[00,02], src=template.tera
    // 07: causal+w/labels: diag_source=01, src=config.yaml
    // 08: both chains: diag_source=00, related=[02,03], src=handler.go
    // 09: nested related: related=[10]; 10→related=[00]
    // 10: nested inner: related=[00]
    // 11: causal+related+labels: diag_source=09, related=[02,04], src=template.tera
    // 12: diamond: diag_source=02, related=[03,04], src=pipeline.py
    // 13: cross-file: related=[00,07,03,10,12] (no own labels, spans via related whose labels cover 5 files)
    // 14: sub-labels: related=[15,16]
    // 15: sub-a: src=handler.go, labels=[PingContext]
    // 16: sub-b: src=handler.go, labels=[database not connected]
    // 17: fan-out: diag_source=11, related=[07,18,19], src=config.yaml
    // 18: fan-sub-a: diag_source=01, src=handler.go
    // 19: fan-sub-b: diag_source=02, src=pipeline.py+log.error
    // 20: full-chaos: diag_source=00, related=[04,07,12,15], src=query.sql
    .{
        .message = "causal chain depth 3 (d04 → d03 → d00)",
        .code = "C004",
        .severity = .Error,
        .help = "each step in the chain adds context",
        .source_idx = 3,
        .labels = &.{prim("TransformError", 479, 14)},
        .diag_source_idx = 3,
    },
    // 05: causal chain depth 4 (d05 → d04 → d03 → d00)
    .{
        .message = "causal chain depth 4 (d05 → d04 → d03 → d00)",
        .code = "C005",
        .severity = .Error,
        .help = "deep causal chains reveal propagation path",
        .source_idx = 3,
        .labels = &.{prim("Semaphore", 269, 9)},
        .diag_source_idx = 4,
    },
    // 06: related with source labels
    .{
        .message = "related with source labels",
        .code = "C006",
        .severity = .Warning,
        .help = "related errors provide additional context",
        .source_idx = 4,
        .labels = &.{ prim("title", 74, 5), sec("items", 107, 5) },
        .related_idxs = &.{ 0, 2 },
    },
    // 07: causal with source labels
    .{
        .message = "causal with source labels",
        .code = "C007",
        .severity = .Error,
        .help = "check the config values",
        .source_idx = 0,
        .labels = &.{ prim("host", 10, 4), sec("8080", 36, 4) },
        .diag_source_idx = 1,
    },
    // 08: both chain types
    .{
        .message = "both chain types (causal + related)",
        .code = "C008",
        .severity = .Error,
        .help = "causal chain explains why, related shows what else",
        .source_idx = 1,
        .labels = &.{prim("HealthCheck", 79, 11)},
        .diag_source_idx = 0,
        .related_idxs = &.{ 2, 3 },
    },
    // 09: nested related (d09 → related[d10], d10 → related[d00])
    .{
        .message = "nested related chain",
        .code = "C009",
        .severity = .Error,
        .help = "nested chains expand recursively",
        .source_idx = 2,
        .labels = &.{prim("ranked AS", 5, 9)},
        .related_idxs = &.{10},
    },
    // 10: nested inner (child of d09, related d00)
    .{
        .message = "nested related inner",
        .code = "C009",
        .severity = .Warning,
        .source_idx = 3,
        .labels = &.{sec("worker", 306, 6)},
        .related_idxs = &.{0},
    },
    // 11: causal + related + labels
    .{
        .message = "causal plus related with labels",
        .code = "C010",
        .severity = .Error,
        .source_idx = 4,
        .labels = &.{prim("for item", 175, 8)},
        .diag_source_idx = 9,
        .related_idxs = &.{ 2, 4 },
    },
    // 12: diamond (one cause, two effects as related)
    .{
        .message = "diamond: one cause with two related effects",
        .code = "C011",
        .severity = .Error,
        .help = "single root cause branches to multiple downstream errors",
        .source_idx = 3,
        .labels = &.{prim("asyncio.gather", 645, 14)},
        .diag_source_idx = 2,
        .related_idxs = &.{ 3, 4 },
    },
    // 13: cross-file (related all have different source files)
    .{
        .message = "cross-file chain spanning all sources",
        .code = "C012",
        .severity = .Error,
        .help = "related errors originate from five different files",
        .related_idxs = &.{ 0, 7, 3, 10, 12 },
    },
    // 14: sub-labels (related each have their own labels on same source)
    .{
        .message = "related with sub-labels",
        .code = "C013",
        .severity = .Warning,
        .help = "each related has its own label within the same source",
        .related_idxs = &.{ 15, 16 },
    },
    // 15: sub-label a
    .{
        .message = "ping context deadline exceeded",
        .code = "C013",
        .severity = .Error,
        .source_idx = 1,
        .labels = &.{prim("PingContext", 273, 11)},
    },
    // 16: sub-label b
    .{
        .message = "database connection is nil",
        .code = "C013",
        .severity = .Error,
        .source_idx = 1,
        .labels = &.{prim("database not connected", 140, 22)},
    },
    // 17: fan-out (one cause, 3 related effects, one effect has sub-cause)
    .{
        .message = "fan-out: one cause with multiple effects",
        .code = "C014",
        .severity = .Error,
        .help = "single failure triggers cascading errors",
        .source_idx = 0,
        .labels = &.{prim("pool_size", 108, 9)},
        .diag_source_idx = 11,
        .related_idxs = &.{ 7, 18, 19 },
    },
    // 18: fan-out sub (related to d17, has own causal source)
    .{
        .message = "timeout connecting to database",
        .code = "C014",
        .severity = .Error,
        .source_idx = 1,
        .labels = &.{prim("fmt.Errorf", 128, 10)},
        .diag_source_idx = 1,
    },
    // 19: fan-out sub (related to d17, has own causal source)
    .{
        .message = "transform pipeline rejected item",
        .code = "C014",
        .severity = .Error,
        .source_idx = 3,
        .labels = &.{prim("log.error", 516, 9)},
        .diag_source_idx = 2,
    },
    // 20: full chaos — all patterns combined
    .{
        .message = "full chaos: causal + related + labels + cross-file + nested",
        .code = "C019",
        .severity = .Error,
        .help = "maximal diagnostic chain complexity",
        .url = "https://example.invalid/chaos",
        .source_idx = 2,
        .labels = &.{ prim("employees", 146, 9), sec("dept_id", 55, 7) },
        .diag_source_idx = 0,
        .related_idxs = &.{ 4, 7, 12, 15 },
    },
};

pub fn build(alloc: std.mem.Allocator, sources: []const diagnostic.NamedSource) ![]diagnostic.Diagnostic {
    const nodes = try alloc.alloc(Node, DESCS.len);
    const diags = try alloc.alloc(diagnostic.Diagnostic, DESCS.len);

    // First pass: create all nodes and their diagnostic wrappers
    for (DESCS, nodes) |desc, *node| {
        const source_code: ?source.SourceCode = if (desc.source_idx) |si|
            sources[si].source()
        else
            null;

        node.* = Node{
            .message = desc.message,
            .code = desc.code,
            .severity = desc.severity,
            .help = desc.help,
            .url = desc.url,
            .source_code = source_code,
            .labels = desc.labels,
            .related = null,
            .diag_source = null,
        };
    }
    for (nodes, diags) |*node, *diag| {
    diag.* = node.implBy();
    }

    var total_related: usize = 0;
    for (DESCS) |desc| total_related += desc.related_idxs.len;

    const related_buf = try alloc.alloc(diagnostic.Diagnostic, total_related);

    {
        var pos: usize = 0;
        for (DESCS) |desc| {
            for (desc.related_idxs) |idx| {
                related_buf[pos] = diags[idx];
                pos += 1;
            }
        }
    }

    // Second pass: wire up chain references
    {
        var pos: usize = 0;
        for (DESCS, nodes) |desc, *node| {
            if (desc.related_idxs.len > 0) {
                node.related = related_buf[pos..][0..desc.related_idxs.len];
                pos += desc.related_idxs.len;
            }
            if (desc.diag_source_idx) |idx| {
                node.diag_source = &diags[idx];
            }
        }
    }

    return diags;
}

const std = @import("std");
